Add Spring Boot Starter for Hazelcast Jet

gurbuzali opened the pull request:
### What this PR does / why do we need it:

Add Spring Boot Starter for Hazelcast Jet
Add health indicator and auto-configuration of this health indicator for Hazelcast Jet


#### Notes to your reviewer:             
See declined PR on spring-boot repository
https://github.com/spring-projects/spring-boot/pull/20518

#### Checklist

- [X] Signed the Hazelcast CLA
- [X] No checkstyle issues (`./gradlew check`)
- [X] No tests failures (`./gradlew test`)
- [X] Documentation which complies with README template (see `templates/README.template.md`).
- [X] Update `List of modules` section on the root README with your module.

eminn commented on jet-spring-boot-starter/src/main/java/com/hazelcast/jet/contrib/actuate/HazelcastJetHealthIndicator.java:
```diff
@@ -0,0 +1,63 @@
+
+    @Override
+    protected void doHealthCheck(Health.Builder builder) {
+        submitJob();
```
Rather than submitting a job to check the status could we get the information from the `LifecycleService` ?

eminn replied:
Also we are reporting always up, right ?

gurbuzali replied:
IMDG integration was trying to do a remote operation to check the availability so I've replicated the logic. But I think we can use the `LifecycleService` here.

If an exception occurs during the job submission then it is caught and reported as down.

eminn commented on settings.gradle:
```diff
@@ -10,6 +10,7 @@ include 'mongodb'
 include 'debezium'
 include 'twitter'
 include 'xa-test'
+include 'jet-spring-boot-starter'
```
I think we should rename module to `spring-boot-starter-jet` to follow the naming convention. See https://github.com/spring-projects/spring-boot/tree/master/spring-boot-project/spring-boot-starters

eminn replied:
Even though community starters do not follow any convention as far as I see.

gurbuzali replied:
there is a convention actually https://docs.spring.io/spring-boot/docs/current/reference/htmlsingle/#using-boot-starter

see the warning in the readme here https://github.com/spring-projects/spring-boot/tree/master/spring-boot-project/spring-boot-starters

```
https://github.com/spring-projects/spring-boot/tree/master/spring-boot-project/spring-boot-starters
```

eminn replied:
Oh, I've missed that. Alright, we are compliant then :)

frant-hartm replied:
Maybe `hazelcast-jet-spring-boot-starter`? _hazelcast jet_ is what ppl will search for the starter.

frant-hartm replied:
I mean in IDE when you fill out the dependency artifact id.

frant-hartm commented on README.md:
```diff
@@ -64,6 +64,11 @@ A Hazelcast Jet connector for consuming data from Twitter stream sources in Jet
 
+### [Jet Spring Boot Starter](jet-spring-boot-starter)
+
+A Spring Boot Starter for Hazelcast Jet which tries to auto-configure Hazelcast
```
tries to auto-configure -> autoconfigures

Don't try, do :-)

frant-hartm commented on jet-spring-boot-starter/gradle.properties:
```diff
@@ -0,0 +1,2 @@
+version = 0.1-SNAPSHOT
```
I know this is a first release, but shouldn't we release the artifact with same version as spring boot we aim to be compatible with? E.g. 2.2.0 for current 2.2.x Spring Boot releases.

gurbuzali commented:
I've addressed the review comments but left the version as is for now. I don't have a concrete opinion about the versioning. what will we do when spring-boot 2.3.x released ? will we just bump the version or create a separate module for for it ? how about the older versions of spring-boot?

@eminn wdyt ?

eminn commented:
I had a look at the community starters and couldn't be able to find any convention there. I think we can start the version with 1.0 which aims for the latest released spring boot version.

eminn commented on README.md:
```diff
@@ -57,17 +59,24 @@ enables Hazelcast Jet pipelines to consume CDC events from various databases.
 Tests to check compatibility of the XA support in your JMS broker or
 JDBC database with Jet's fault tolerance.
 
+### [Jet Spring Boot Starter](jet-spring-boot-starter)
```
`hazelcast-jet-spring-boot-starter` ?

eminn commented on hazelcast-jet-spring-boot-starter/README.md:
````diff
@@ -0,0 +1,105 @@
+```
+<dependency>
+    <groupId>com.hazelcast.jet.contrib</groupId>
+    <artifactId>jet-spring-boot-starter</artifactId>
````
Can you match this with the module name ?

eminn commented on hazelcast-jet-spring-boot-starter/README.md:
````diff
@@ -0,0 +1,105 @@
+
+or if you are using Gradle: 
+```
+compile group: 'com.hazelcast.jet.contrib', name: 'jet-spring-boot-starter', version: ${version}
````
same here

eminn commented on hazelcast-jet-spring-boot-starter/src/test/java/com/hazelcast/jet/contrib/actuate/HazelcastJetHealthIndicatorTests.java:
```diff
@@ -0,0 +1,77 @@
+    private final JetInstance jet = mock(JetInstance.class);
+
+    @Test
+    public void hazelcastJetUp() {
```
Can we rename this to conform with given, when, then template ?

eminn commented on hazelcast-jet-spring-boot-starter/src/test/java/com/hazelcast/jet/contrib/actuate/HazelcastJetHealthIndicatorTests.java:
```diff
@@ -0,0 +1,77 @@
+    }
+
+    @Test
+    public void hazelcastJetDown() {
```
same here
